### PR TITLE
Add digest authentication

### LINKED ISSUE #572: Does resty support digest authentication
Does resty support digest authentication or is there any plan for it

### PR DESCRIPTION
**Pull request by segevda**
- Adhere to RFC 7616 ("HTTP Digest Access Authentication")
- Added SetDigestAuth methods for Client and Request
- Currently not supporting auth-int Quality of Protection

closes #467
closes #572

### CODE REVIEW

File: digest.go [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+		sl = append(sl, fmt.Sprintf("nc=%08x", c.nc))
+		sl = append(sl, fmt.Sprintf(`cnonce="%s"`, c.cNonce))
+	}
+
+	return fmt.Sprintf("Digest %s", strings.Join(sl, ", ")), nil
+}
+
+func (c *credentials) validateQop() error {
+	// Currently only supporting auth quality of protection. TODO: add auth-int support
+	// NOTE: cURL support auth-int qop for requests other than POST and PUT (i.e. w/o body) by hashing an empty string
```
Comment (jeevatkm) [inline digest.go:216]: @segevda auth-int can be done later as a separate PR.

File: digest.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
+package resty
```
Comment (jeevatkm) [inline digest.go:1]: @segevda Did you use any code reference or documentation for this implementation? If yes, please add those references in the code file.

**Review by jeevatkm — CHANGES_REQUESTED** [review]
@segevda Thank you of your PR and contribution. I'm sorry for the delayed attention on the PR.
Can you please check the review comments?

File: digest.go (hunk 1 above)
Comment (segevda) [reply digest.go:216]: Yes I agree. I think this comment should remain for reference. The cURL implementation is something worth considering

File: digest.go (hunk 2 above)
Comment (segevda) [reply digest.go:1]: This code was loosely inspired by [bobziuchkovski/digest](https://github.com/bobziuchkovski/digest). I don't mind referencing it, just not sure if this has any license implications. The code in that repo is under Apache 2.0 and resty is under MIT. WDYT?

Comment (jeevatkm) [reply digest.go:1]: I see, thanks for the info.

**Review by jeevatkm — APPROVED** [review]
@segevda Thanks for the PR.

**Comment by segevda** [comment]
> @segevda It seems test cases coverage check is failing, can you please check it?

@jeevatkm just to make sure I understand, the failure here is because the total codebase coverage dropped due to this addition?

**Comment by jeevatkm** [comment]
> > @segevda It seems test cases coverage check is failing, can you please check it?
> 
> @jeevatkm just to make sure I understand, the failure here is because the total codebase coverage dropped due to this addition?

Yes, @segevda.

**Comment by segevda** [comment]
@jeevatkm I don't see how I can improve the coverage on this file any more. All missing lines are error checks for std lib functions. Please assist.

**Comment by jeevatkm** [comment]
@segevda I see, I believe you have looked at this page. Which lists those return error lines, correct?
https://github.com/go-resty/resty/pull/583/checks?check_run_id=11833912592 

Possibly, we could supply incorrect input values to produce the error in the test! what are you thoughts?

**Comment by segevda** [comment]
@jeevatkm I found a way to reduce error checking and thus to improve coverage:
Hash.Write calls are documented to never return an error ([see here](https://pkg.go.dev/hash#Hash)).
Also, changed the way we assign the digest Transport to avoid data races.

Please review :pray:

File: digest.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+	"strings"
+)
+
+var (
+	ErrBadChallenge    = errors.New("challenge is bad")
+	ErrCharset         = errors.New("unsupported charset")
+	ErrAlgNotSupported = errors.New("algorithm is not supported")
+	ErrQopNotSupported = errors.New("no supported qop in list")
+	ErrNoQop           = errors.New("qop must be specified")
+)
```
Comment (jeevatkm) [inline digest.go:22]: @segevda Can you please update Error names to following?
```go
var (
	ErrDigestBadChallenge    = errors.New("digest: challenge is bad")
	ErrDigestCharset         = errors.New("digest: unsupported charset")
	ErrDigestAlgNotSupported = errors.New("digest: algorithm is not supported")
	ErrDigestQopNotSupported = errors.New("digest: no supported qop in list")
	ErrDigestNoQop           = errors.New("digest: qop must be specified")
)
```

**Review by jeevatkm — CHANGES_REQUESTED** [review]
@segevda Thanks for improving the PR in the last iteration. I have one last comment, can you have a look?

File: digest.go (hunk 3 above)
Comment (segevda) [reply digest.go:22]: Done

**Review by jeevatkm — APPROVED** [review]
Thanks @segevda for resolving the comments

**Pull request merged**